src/sage/modules/free_module.py

```diff
@@ -7601,7 +7668,11 @@ def echelonized_basis(self):
             sage: V.span(B) == M
             True
         """
-        return self.__echelonized_basis
+        C = self.element_class
+        echelonized_basis = self.echelonized_basis_matrix().rows()
+        w = [C(self, x.list(), coerce=False, copy=True)
+                for x in echelonized_basis]
```

Comment:
```suggestion
        echelonized_basis = self.echelonized_basis_matrix().rows()
        if echelonized_basis and echelonized_basis[0].parent() is not self:
            w = [self(x) for x in echelonized_basis]
       else:
            w = echelonized_basis
```
Let's avoid recreating the vectors when they are already in the correct parent. However, I would expect the conversion map (in particular, retracting from the ambient space) to be faster than getting the list. In particular, if the vectors are big and sparse, this should (in principle) keep them sparse. Perhaps this would lead to infinite loops? Any thoughts about this?

Reply:
at least the tests in this file pass.

Yes, I think the conversion map is slower by at most O(1)… in principle.

in the file:

```
    def _element_constructor_(self, x, coerce=True, copy=True, check=True):
        r"""
        Create an element of this module from ``x``.

        The ``coerce`` and ``copy`` arguments are passed on to the underlying
        element constructor.

        EXAMPLES::

            sage: S.<x,y,z> = PolynomialRing(QQ)
            sage: M = S**2
            sage: N = M.submodule([vector([x - y, z]), vector([y*z , x*z])])
            sage: Q = M.quotient_module(N)
            sage: Q(0)
            (0, 0)
            sage: Q([x, x + y])
            (x, x + y)
            sage: phi = Q.coerce_map_from(M)
            sage: phi(M.gen(1))
            (0, 1)
            sage: _ in Q
            True
        """
        if isinstance(x, (int, sage.rings.integer.Integer)) and x == 0:
            return self.zero_vector()
        elif isinstance(x, FreeModuleElement):
            if x.parent() is self:
                if copy:
                    return x.__copy__()
                else:
                    return x
            x = x.list()
⋮
```

it could be faster if some subclass override `_element_constructor_` to avoid the copy or do the copy in C, but this sounds like a lot of work to check. likely out of scope of this pull request.